shared/lib_irradproc.cpp

```diff
@@ -219,30 +219,46 @@ void solarpos(int year,int month,int day,int hour,double minute,double lat,doubl
 		E = E - 24.0;
 
 	arg = -tan(lat)*tan(dec);
-	if (arg >= 1.0)  /* No sunrise, continuous nights */
+	if (arg >= 1.0)  // No sunrise, continuous nights
 	{
 		ws = 0.0;                        
-		sunrise = 24;
-		sunset = 0;
+		sunrise = 25.0;
+		sunset = -1.0;
 	}
-	else if (arg <= -1.0) /* No sunset, continuous days */
+	else if (arg <= -1.0) // No sunset, continuous days
 	{
 		ws = M_PI;                          
-		sunrise = 0;
-		sunset = 24;
+		sunrise = -1.0;
+		sunset = 25.0;
 	}
 	else
 	{
-		ws = acos(arg); /* Sunrise hour angle in radians */
-		/* Sunrise and sunset in local standard time */
+		ws = acos(arg); // Sunrise hour angle in radians
+		// Sunrise and sunset in local standard time
 		sunrise = 12.0 - (ws / DTOR) / 15.0 - (lng / 15.0 - tz) - E; //sunrise in units of hours (e.g. 5.25 = 5:15 am)
 		sunset = 12.0 + (ws / DTOR) / 15.0 - (lng / 15.0 - tz) - E; //sunset in units of hours (e.g. 18.75 = 6:45 pm)
-		//if time zone and longitude are opposite signs (happens near international dateline and greenwich meridian), then sometimes sunrise/sunset will be off by 24 hrs
-		//add catches to check for that and correct it if so
-		if (sunrise > 24) sunrise -= 24;
-		else if (sunrise < 0) sunrise += 24;
-		if (sunset > 24) sunset -= 24;
-		else if (sunset < 0) sunset += 24;
+		//now a bunch of error checks to try to correctly catch weird behavior
+		//both sunrise and sunset may be shifted by 24 hours (example: Fiji- positive tz negative lng), if so, roll them both back
+		if (sunrise > 24.0 && sunset > 24.0)
+		{
+			sunrise -= 24.0;
+			sunset -= 24.0;
+		}
+		/*//if only sunset is greater than 24, it actually sets the next day, but ws not quite big enough to be caught (example: Point Hope, AK)
```

Comment:
It seems like there are 4 possible odd cases:
sunrise > 24 and sunset > 24
sunrise < 0 and sunset < 0
sunrise > 24 and sunset < 0
sunrise < 0 and sunset > 24
The last two seem to be handled in lines 222-232 whereas the first two inside 234-262, but is it possible that |arg| < 1.0 but we still end up in the last two cases?

Reply:
The logic here is that in case 1 or case 2, the equations just aren't working for a weird longitude/time zone combination (which is what we've found with test cases historically), so they legitimately both just need to be rolled by 24 hours. I don't think Case 3 is possible without it being caught by the |arg| < 1.0 check (or something going VERY wrong with the inputs), because sunrise= 12-x and sunset=12+x. Case 4 I would say is two separate cases: sunrise < 0 OR sunset > 24. Because cases 1 and 2 have hopefully ruled out cases of the equations not working for a weird tz/long combo, then likely Case 4A and 4B are legitimate sunrises/sunsets that occur on a different day in high latitude locations. Since they are legit, we want solarpos to return these times as they are, and have subsequent logic able to deal with them (lines 1140-1156).

Reply:
Ok, thanks for clarifying!